File: packages/anki/sync/engine.py
"""Sync engine orchestrator."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

from packages.anki.sync.progress import ProgressTracker, SyncPhase
from packages.common.logging import get_logger

if TYPE_CHECKING:
    from packages.anki.sync.state import StateDB

logger = get_logger(module=__name__)


@dataclass(frozen=True, slots=True)
class SyncResult:
    """Result of a sync operation."""

    cards_created: int = 0
    cards_updated: int = 0
    cards_deleted: int = 0
    cards_skipped: int = 0
    errors: int = 0
    duration_ms: int = 0


class SyncEngine:
    """Orchestrate sync lifecycle.

    The engine manages phases of a sync operation, delegating actual
    card operations to injected components via the state database.
    """

    def __init__(
        self,
        state_db: StateDB,
        *,
        progress: ProgressTracker | None = None,
    ) -> None:
        self._state_db = state_db
        self._progress = progress or ProgressTracker()
        self._cards_created = 0
        self._cards_updated = 0
        self._cards_deleted = 0
        self._cards_skipped = 0
        self._errors = 0

    @property
    def state_db(self) -> StateDB:
        """Access the state database."""
        return self._state_db

    @property
    def progress(self) -> ProgressTracker:
        """Access the progress tracker."""
        return self._progress

    def sync(self, *, dry_run: bool = False) -> SyncResult:
        """Run sync lifecycle.

        Args:
            dry_run: If True, scan but do not apply changes.

        Returns:
            SyncResult with operation counts.
        """
        start = time.monotonic()
        session = self._progress.snapshot().session_id
        logger.info("sync.started", session_id=session, dry_run=dry_run)

        try:
            self._progress.set_phase(SyncPhase.SCANNING)
            self._scan()

            if not dry_run:
                self._progress.set_phase(SyncPhase.APPLYING)
                self._apply()

            self._progress.complete(success=True)
        except Exception:
            self._progress.complete(success=False)
            self._errors += 1
            logger.exception("sync.failed", session_id=session)
            raise

        elapsed_ms = int((time.monotonic() - start) * 1000)
        result = SyncResult(
            cards_created=self._cards_created,
            cards_updated=self._cards_updated,
            cards_deleted=self._cards_deleted,
            cards_skipped=self._cards_skipped,
            errors=self._errors,
            duration_ms=elapsed_ms,
        )
        logger.info("sync.completed", session_id=session, result=result)
        return result

    def _scan(self) -> None:
        """Scan for changes (subclasses or callers populate state)."""
        existing = self._state_db.get_all()
        self._progress.set_total(len(existing))
        logger.debug("sync.scan_complete", card_count=len(existing))

    def _apply(self) -> None:
        """Apply changes (subclasses override for actual operations)."""
        logger.debug("sync.apply_complete")

File: crates/anki-sync/src/engine.rs
use std::time::Instant;

use crate::progress::{ProgressTracker, SyncPhase};
use crate::state::{StateDB, StateDbError};

/// Result of an engine-level sync operation.
#[derive(Debug, Clone)]
pub struct SyncResult {
    pub cards_created: i32,
    pub cards_updated: i32,
    pub cards_deleted: i32,
    pub cards_skipped: i32,
    pub errors: i32,
    pub duration_ms: i64,
}

/// Phased sync engine with pluggable state and progress tracking.
///
/// Lifecycle: INITIALIZING -> SCANNING -> APPLYING -> COMPLETED/FAILED
pub struct SyncEngine {
    state_db: StateDB,
    progress: ProgressTracker,
}

impl SyncEngine {
    pub fn new(state_db: StateDB, progress: Option<ProgressTracker>) -> Self {
        let progress = progress.unwrap_or_else(|| ProgressTracker::new(None));
        Self { state_db, progress }
    }

    /// Access the state database.
    pub fn state_db(&self) -> &StateDB {
        &self.state_db
    }

    /// Access the progress tracker.
    pub fn progress(&self) -> &ProgressTracker {
        &self.progress
    }

    /// Run the sync lifecycle.
    pub fn sync(&mut self, dry_run: bool) -> Result<SyncResult, StateDbError> {
        let start = Instant::now();

        // SCANNING phase
        self.progress.set_phase(SyncPhase::Scanning);
        let states = self.state_db.get_all()?;
        self.progress.set_total(states.len() as i32);

        // APPLYING phase (skip for dry run)
        if !dry_run {
            self.progress.set_phase(SyncPhase::Applying);
        }

        // COMPLETED
        self.progress.set_phase(SyncPhase::Completed);

        let duration_ms = start.elapsed().as_millis() as i64;

        Ok(SyncResult {
            cards_created: 0,
            cards_updated: 0,
            cards_deleted: 0,
            cards_skipped: 0,
            errors: 0,
            duration_ms,
        })
    }
}
